server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java

```diff
@@ -121,28 +207,189 @@ public ClusterState closeIndices(ClusterState currentState, final Index[] indice
+            }
+            blockedIndices.add(index);
         }
 
-        ClusterState updatedState = ClusterState.builder(currentState).metaData(mdBuilder).blocks(blocksBuilder).build();
+        logger.debug(() -> new ParameterizedMessage("adding block to indices {}",
+            blockedIndices.stream().map(Object::toString).collect(Collectors.joining(","))));
+        return ClusterState.builder(currentState).blocks(blocks).metaData(metadata).routingTable(routingTable.build()).build();
+    }
 
-        RoutingTable.Builder rtBuilder = RoutingTable.builder(currentState.routingTable());
-        for (IndexMetaData index : indicesToClose) {
-            rtBuilder.remove(index.getIndex().getName());
+    /**
+     * Step 2 - Wait for indices to be ready for closing
+     * <p>
+     * This step iterates over the indices previously blocked and sends a {@link TransportVerifyShardBeforeCloseAction} to each shard. If
+     * this action succeed then the shard is considered to be ready for closing. When all shards of a given index are ready for closing,
+     * the index is considered ready to be closed.
+     */
+    class WaitForClosedBlocksApplied extends AbstractRunnable {
+
+        private final Set<Index> blockedIndices;
+        private final @Nullable TimeValue timeout;
+        private final ActionListener<Map<Index, AcknowledgedResponse>> listener;
+
+        private WaitForClosedBlocksApplied(final Set<Index> blockedIndices,
+                                           final @Nullable TimeValue timeout,
+                                           final ActionListener<Map<Index, AcknowledgedResponse>> listener) {
+            if (blockedIndices == null || blockedIndices.isEmpty()) {
+                throw new IllegalArgumentException("Cannot wait for closed block to be applied to null or empty list of blocked indices");
+            }
+            this.blockedIndices = blockedIndices;
+            this.listener = listener;
+            this.timeout = timeout;
+        }
+
+        @Override
+        public void onFailure(final Exception e) {
+            listener.onFailure(e);
+        }
+
+        @Override
+        protected void doRun() throws Exception {
+            final Map<Index, AcknowledgedResponse> results = ConcurrentCollections.newConcurrentMap();
+            final CountDown countDown = new CountDown(blockedIndices.size());
+            final ClusterState state = clusterService.state();
+            for (Index blockedIndex : blockedIndices) {
+                waitForShardsReadyForClosing(blockedIndex, state, timeout, new ActionListener<AcknowledgedResponse>() {
+                    @Override
+                    public void onResponse(final AcknowledgedResponse result) {
+                        results.put(blockedIndex, result);
+                        processIfFinished();
+                    }
+
+                    @Override
+                    public void onFailure(final Exception e) {
```

Comment:
is this ever called in waitForShardsReadyForClosing?

Reply:
It's not, you're right. I changed for a simple Consumer

Reply:
Done